docs/introduction-to-matching-logic.tex

```diff
@@ -0,0 +1,2767 @@
+iff $\rho(x) \not\in \overline{\rho}(\varphi)$, and
+$\overline{\rho}(x \in_{s_1}^{s_2} \varphi) = M_{s_2}$
+iff $\rho(x) \in \overline{\rho}(\varphi)$.
+It is convenient to extend the definition of membership to cases when the first
+component is not a variable, but a pattern:
+if $\psi, \varphi \in \Pattern_{s_1}$, and $s_1,s_2\in S$,
+then let
+$$
+\begin{array}{@{}rcll}
+\psi \in_{s_1}^{s_2} \varphi & \ \ \ \ \ \equiv \ \ \ \ \ &
+\lceil \psi \wedge \varphi\rceil_{s_1}^{s_2}
+& % \ \ \ \ \ \mbox{where } x \in \Var_{s_1},\varphi\in \Pattern_{s_1}\\
+\end{array}
+$$
+This allows a uniform interface for pattern constructs, which yields neat
+implementation of operations such as substitution.
+As an example, consider $\varphi_1[\varphi_2 / x]$ where we
+substitute $x$ for $\varphi_2$ in $\varphi_1$.
+If the membership construct only accept variables as its first component, then
+the substitution is only defined when $x$ does not occur free in the first
+component of any membership construct in $\varphi_1$, and any implementation of
+substitution will have to take that into account and perform an
+``occurs check'' for $x$ in $\varphi_1$, and this could be expensive.
+
+When writing patterns, the following precedence and associativity rules are
+useful to reduce the
+number of necessary parentheses.
+\begin{center}
+ \begin{tabular}{c|c|c}
+  Connective      &     Precedence & Associativity \\\hline
+  $\neg$          &     1          &  --           \\\hline
+  $=$, $\in$      &     2          &  --           \\\hline
+  $\wedge$        &     3          &  left         \\\hline
+  $\vee$          &     4          &  left         \\\hline
+  $\to$           &     5          &  right        \\\hline
+  $\leftrightarrow$ &   6          &  --           \\\hline
+  $\exists$,$\forall$ & 7          &  --           \\\hline
+ \end{tabular}
+\end{center}
+As a convention, binders ($\forall$ and $\exists$) have a lower precedence than
+other connectives.
+This means that the scope of a binder extends as far to the right as possible.
+
+Since $s_1$ and $s_2$ can usually be inferred from context,
+we write $\lceil\_\rceil$ or $\lfloor\_\rfloor$ instead of
+$\lceil\_\rceil_{s_1}^{s_2}$ or $\lfloor\_\rfloor_{s_1}^{s_2}$, respectively,
+and similarly for the equality and membership.
+Also, if the sort decorations cannot be inferred from context, then we assume
+the stated property/axiom/rule holds for all such sorts.
+\info{Refer to this later when we talk about parameters.}
+%
+For example, the generic pattern axiom ``$\lceil x \rceil$ where $x\in\Var$''
+replaces all the axioms $\lceil x\!:\!s_1 \rceil_{s_1}^{s_2}$ above for all
+the definedness symbols for all the sorts $s_1$ and $s_2$.
+
+Note that, by default, symbols are interpreted as relations in matching logic.
+It is often the case, though, that we want symbols to be interpreted as
+\emph{functions}.
+This can be easily done by axiomatically constraining those symbols to evaluate to
```

Comment:
I think it can be confusing that _function_ in this context means evaluating to _exactly_ one value, while the glossary describes it as evaluating to _at most_ one value.